### ISSUE TITLE
[iOS 13] callKit displayed from AppDelegate.m before incoming call listeners

### ISSUE THREAD

**Issue by ahmadworks** [react-native-webrtc/react-native-callkeep#107]
After updating to 3.0.3 and updating my code to meet iOS 13 requirements so CallKit is reported from AppDelegate. But the problem is that when app is closed or in background, the callKit displays very fast before my call class get loaded. So the callKit is shown on screen before listeners are ready and if I accept or reject the call immediately nothing happened. I need to wait for 3 seconds until call listeners are ready then I can accept or reject.

How can I delay the callKit in AppDelegate until my call class is fully loaded?

**Comment by manuquentin** [comment]
Hi @ahmadworks.
It's the normal behaviour since in ios 13 we should report all calls coming from a VoIP push.
Please refer to the [doc about it](https://github.com/react-native-webrtc/react-native-callkeep#pushkit) and the [corresponding PR](https://github.com/react-native-webrtc/react-native-callkeep/pull/86).

**Comment by ahmadworks** [comment]
I don't think this should be the normal behaviour even after iOS 13, I can report to call but the call should appear on screen only after the app knows how to deal with the call. in current case what should happen if user quickly tapped answer before the app knows how it should handle the answer and reject.

**Comment by danjenkins** [comment]
@ahmadworks this is needed for ios13 so we need to think collectively about how to work around that issue. I think from my perspective we need the native side to know that the JS bridge isn't up yet and so wait until it is and then emit the event of "yes someone accepted the call"

**Comment by Prussich** [comment]
@danjenkins  also to emit UUID data of active CallKit call (for example, to update the incoming call display). Because RN doesn't know the current UUID call after the bridge is up. didDisplayIncomingCall not called (error "Sending 'RNCallKeepDidDisplayIncomingCall' with no listeners registered.)

**Comment by manuquentin** [comment]
@Prussich you have to listen to the events at the very beginning of the loading of your app (eg. in the `index.ios.js`). You'll next have to wait for you WebRTC call to be set up before really answer it.

**Comment by bmp123** [comment]
Is it possible to implement dnd mode with RNCallKeep? how can I use rejectCall if receive voip push?

**Comment by danjenkins** [comment]
@bmp123 please create a new issue for that :)

**Comment by Prussich** [comment]
@manuquentin 
As I see it:
1) App wake up after receive VoIP-Push.
2) App calling reportNewIncomingCall from VoIP-push delegate (as quickly, as Apple asks)
```
...
NSString *uuid = [[[NSUUID UUID] UUIDString] lowercaseString];
...
[RNCallKeep reportNewIncomingCall:uuid handle:handle handleType:@"generic" hasVideo:false localizedCallerName:callerName fromPushKit: YES];
```
RN bridge still not live. And there are no CallKeep listeners yet.
3) RN bridge is started. The RN part with CallKeep listeners starts up. 

But, RNCallKeepDidDisplayIncomingCall was not called and RNCallKeep doesn't know which uuid was used in reportNewIncomingCall:uuid in native part.

Please correct me if everything is wrong and you can know in the RN part which UUID of the incoming call was used before in the native part of the application.

**Comment by Prussich** [comment]
@manuquentin I also try to listen to the events at the very beginning of the loading of my app:

At the top of index.ios.js:
```
const onAnswerCallAction = ({callUUID}) => {
    RNFS.writeFile(path, '\n[' + moment().locale('en').format('YYYY-MM-DD HH:mm:ss') + ']: ' + '[index]: onAnswerCallAction from App start\n', 'utf8');
};

const onIncomingCallDisplayed = ({error, callUUID, handle, localizedCallerName, fromPushKit}) => {
    RNFS.writeFile(path, '\n[' + moment().locale('en').format('YYYY-MM-DD HH:mm:ss') + ']: ' + '[index]: onIncomingCallDisplayed from App start:' + callUUID + '\n', 'utf8');
};

RNCallKeep.setup(options);
RNCallKeep.addEventListener('answerCall', onAnswerCallAction);
RNCallKeep.addEventListener('didDisplayIncomingCall', onIncomingCallDisplayed);
```

and I don't see any event records in log file about "onIncomingCallDisplayed". In the native part, sending the onIncomingCallDisplayed event fires earlier than there is a subscription to the event in JS. Any ideas how to fix this?

**Comment by manuquentin** [comment]
@Prussich there is a timeout to work around this issue [here](https://github.com/react-native-webrtc/react-native-callkeep/blob/master/ios/RNCallKeep/RNCallKeep.m#L19).
The app will wait 5s in release and 10s in debug before [sending the event](https://github.com/react-native-webrtc/react-native-callkeep/blob/master/ios/RNCallKeep/RNCallKeep.m#L525).

Here the event triggered too soon event with this workaround ?

**Comment by Prussich** [comment]
Good idea! But it work around for the **outgoing call** (RNCallKeepDidReceiveStartCallAction event).
I 'll try to make the same implementation with the timeout for **incoming calls** for the RNCallKeepDidDisplayIncomingCall event in RNCallKeep.m.

**Comment by pravinjohn** [comment]
Hello, please do share if anyone got a solution for this. Thanks!

**Comment by pravinjohn** [comment]
The following seems to be working.
1. Make sure it set callkeep and call related (voximplant/webrtc in my case) event listeners as soon as the app loads.
2. In RNCallKeep.m:
2.a. Delay --> sendEventWithName:RNCallKeepDidDisplayIncomingCall
`callInDateAndTime = [NSDate date];
[RNCallKeep initCallKitProvider];
    int delayInSeconds2 = 5;
    [sharedProvider reportNewIncomingCallWithUUID:uuid update:callUpdate completion:^(NSError * _Nullable error) {

        dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds2 * NSEC_PER_SEC));
        dispatch_after(popTime, dispatch_get_main_queue(), ^{
            RNCallKeep *callKeep = [RNCallKeep allocWithZone: nil];
            [callKeep sendEventWithName:RNCallKeepDidDisplayIncomingCall body:@{ @"error": error ? error.localizedDescription : @"", @"callUUID": uuidString, @"handle": handle, @"localizedCallerName": localizedCallerName, @"fromPushKit": fromPushKit ? @"1" : @"0" }];
            if (error == nil) {
                // Workaround per https://forums.developer.apple.com/message/169511
                if ([callKeep lessThanIos10_2]) {
                    [callKeep configureAudioSession];
                }
            }
        });

    }];`
2.b. Delay --> sendEventWithName:RNCallKeepPerformAnswerCallAction
`int delayInSeconds3 = 0;
    if (callInDateAndTime){
        NSDate *currentDateAndTime = [NSDate date];
        NSTimeInterval elapsed = [currentDateAndTime timeIntervalSinceDate:callInDateAndTime];
        if (elapsed <= 1){
            delayInSeconds3 = 1;
        }
    }
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds3 * NSEC_PER_SEC));
    dispatch_after(popTime, dispatch_get_main_queue(), ^{
        [self configureAudioSession];
        [self sendEventWithName:RNCallKeepPerformAnswerCallAction body:@{ @"callUUID": [action.callUUID.UUIDString lowercaseString] }];
        [action fulfill];
    });`
2.c. Delay --> sendEventWithName:RNCallKeepPerformEndCallAction
`int delayInSeconds4 = 0;
    if (callInDateAndTime){
        NSDate *currentDateAndTime = [NSDate date];
        NSTimeInterval elapsed = [currentDateAndTime timeIntervalSinceDate:callInDateAndTime];
        if (elapsed <= 1){
            delayInSeconds4 = 1;
        }
    }
    dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds4 * NSEC_PER_SEC));
    dispatch_after(popTime, dispatch_get_main_queue(), ^{
        [self sendEventWithName:RNCallKeepPerformEndCallAction body:@{ @"callUUID": [action.callUUID.UUIDString lowercaseString] }];
        [action fulfill];
    });`
3. Most importantly, I have no prior ios experience. Take this with a bucket of salt and please post a proper solution if you have one. Thanks.

**Comment by danwhite-ad** [comment]
Would it not be possible create a method `AnswerCall(UUID)` which calls `[action fulfill];` currently called within `performAnswerCallAction`, this way the app should show connecting until your app has managed to reconnect and retrieve the call and actually answer the call. I am not that familiar with native iOS code otherwise I would attempt to see if the concept is possible.

**Comment by konradkierus** [comment]
For anyone interested - please check my proof of concept to solve this issue
https://github.com/softwarehutpl/react-native-callkeep/commit/b7df3b4863f1bcefd8fd89fcdaf7bcd10dff207e

The idea is to gather events which should be sent through JS bridge if it isn't ready. Once JS part is ready then send didLoadWithEvents with all gathered events.
Next, on JS part, events needs to be handled depending of the project needs. For example, you may be interested in checking of only last "end call" action but it's worth having "did display call" action as well in order to get all push/call-related data.

Disclaimer:
1. It wasn't fully tested yet.
2. It may not be necessary to wrap all events.
3. It may require some examples in README of how to use didLoadWithEvents event.

**Comment by manuquentin** [comment]
Thanks @konradkierus,
Can you make a PR with these changes ? I'll be happy to test the,.

**Comment by konradkierus** [comment]
@manuquentin sure, please check #169

**Comment by mariouzae** [comment]
I see sometimes the JS Bridge doesn't start after receive a VoIP PN. It receives the VoIP PN, I report the new incoming call but the app never starts. It seems the JS Bridge thread is paused until you take some action on the CallKit screen.

Even if I wait for a huge time, I don't see the app starting.

Does anyone have this issue too?

**Comment by zxcpoiu** [comment]
@mariouzae 

You can try NOT calling `completion()` first. If you marked `completion()` and you've see your js bridge worked, then you might encounter the similar issue.

https://github.com/react-native-webrtc/react-native-voip-push-notification/pull/52

This PR is a try which aims to solve `calling completion() too early to stop JS initialization` issue

**Comment by manuquentin** [comment]
Should be fixed in #205

**Comment by danjenkins** [comment]
A lot of work has gone into this now.... closing and if people still have large issues... then they should re-open a new issue

**Comment by danwhite-ipc** [comment]
Do you know when #205 is due to be merged?

**Comment by danjenkins** [comment]
@danwhite-ipc nope - have you tried running it and seeing if it helps solve your problem?
